Frameworks/CoreText/DWriteWrapper.mm

```diff
@@ -0,0 +1,392 @@
+        return S_OK;
+    };
+
+    HRESULT STDMETHODCALLTYPE IsPixelSnappingDisabled(_In_opt_ void* clientDrawingContext, _Out_ BOOL* isDisabled) throw() {
+        *isDisabled = false;
+        return S_OK;
+    };
+
+    HRESULT STDMETHODCALLTYPE GetCurrentTransform(_In_opt_ void* clientDrawingContext, _Out_ DWRITE_MATRIX* transform) throw() {
+        return S_OK;
+    };
+
+    HRESULT STDMETHODCALLTYPE GetPixelsPerDip(_In_opt_ void* clientDrawingContext, _Out_ float* pixelsPerDip) throw() {
+        *pixelsPerDip = 1.0;
+        return S_OK;
+    };
+
+private:
+    ComPtr<IDWriteFactory> _dwriteFactory;
+};
+
+CustomTargetTextRenderer::CustomTargetTextRenderer() {
+    _dwriteFactory = GetDWriteFactoryInstance();
+}
+
+HRESULT CustomTargetTextRenderer::RuntimeClassInitialize() {
+    _dwriteFactory = GetDWriteFactoryInstance();
+    return S_OK;
+}
+
+/**
+ * Helper method that will create a TextLayout and TextRenderer object with the given _CTTypesetter object, string range and frame size
+ * to render to.
+ *
+ * @parameter ts _CTTypesetter object.
+ * @parameter range string range to consider for rendering.
+ * @parameter frameSize frame constrains to render the text on.
+ * @parameter glyphDetails pointer to the _DWriteGlyphRunDetails object that contains the glyph run details that was rendeered.
+ */
+static void _callCustomRender(_CTTypesetter* ts, CFRange range, CGRect frameSize, _DWriteGlyphRunDetails* glyphDetails) {
+    ComPtr<IDWriteTextLayout> textLayout = _createDWriteTextLayout(ts, range, frameSize);
+    ComPtr<CustomTargetTextRenderer> textRenderer = Make<CustomTargetTextRenderer>();
+    textLayout->Draw(glyphDetails, textRenderer.Get(), 0, 0);
+}
+
+/**
+ * Helper method to retrieve font fmaly names installed in the system.
+ *
+ * @return Unmutable array of font family name strings that are installed in the system.
+ */
+static NSArray<NSString*>* _dwriteGetFamilyNames() {
+    NSMutableArray<NSString*>* fontFamilyNames = [NSMutableArray<NSString*> array];
+
+    // Get the direct write factory instance
+    ComPtr<IDWriteFactory> dwriteFactory = GetDWriteFactoryInstance();
+
+    // Get the system font collection.
+    ComPtr<IDWriteFontCollection> fontCollection;
+    THROW_IF_FAILED(dwriteFactory->GetSystemFontCollection(&fontCollection));
+
+    // Get the number of font families in the collection.
+    uint32_t count = 0;
+    count = fontCollection->GetFontFamilyCount();
+
+    for (uint32_t i = 0; i < count; ++i) {
+        // Get the font family.
+        ComPtr<IDWriteFontFamily> fontFamily;
+        THROW_IF_FAILED(fontCollection->GetFontFamily(i, &fontFamily));
+
+        // Get a list of localized strings for the family name.
+        ComPtr<IDWriteLocalizedStrings> familyNames;
+        THROW_IF_FAILED(fontFamily->GetFamilyNames(&familyNames));
+
+        [fontFamilyNames addObject:_convertLocalizedStringToNSString(familyNames.Get())];
+    }
```

Comment:
We both ended up writing this helper actually - I'll defer to @mnithish 's since I like the placement better, and because mine is messier due to trying to stick to CF functions at the time.

---
In reply to: [77756735](https://github.com/Microsoft/WinObjC/pull/950#discussion_r77756735) [](ancestors = 77756735)